File: 2015/day_02_2015.py
# day_02_2015.py - https://adventofcode.com/2015/day/2


import os
from pathlib import Path


def paper_and_ribbon(dim):
    l, w, h = sorted([int(i) for i in dim.split("x")])
    a, b, c = [(l * w), (w * h), (h * l)]

    surface_area = (2 * a) + (2 * b) + (2 * c)
    paper = surface_area + min(a, b, c)

    wrap = 2 * l + 2 * w
    bow = l * w * h
    ribbon = wrap + bow

    return (paper, ribbon)


def main():
    os.chdir(Path(__file__).parent)
    with open("day_02_2015_input.txt", "r") as infile:
        data = [line.strip() for line in infile.readlines()]

    total_paper, total_ribbon = 0, 0
    for gift in data:
        paper, ribbon = paper_and_ribbon(gift)
        total_paper += paper
        total_ribbon += ribbon

    print(total_paper)
    print(total_ribbon)


if __name__ == "__main__":
    main()

File: 2015/javascript/day_02_2015.js
const fs = require("fs");
const data = fs.readFileSync("day_02_2015_input.txt", "utf-8").split("\r\n");

let totalPaper = 0;
let totalRibbon = 0;

const paperAndRibbon = (dims) => {
  dims = dims.split("x");
  dims = [Number(dims[0]), Number(dims[1]), Number(dims[2])];
  let [l, w, h] = dims.sort(function (x, y) {
    return x - y;
  });
  let [a, b, c] = [l * w, w * h, h * l];

  let surfaceArea = 2 * a + 2 * b + 2 * c;
  let paper = surfaceArea + Math.min(a, b, c);

  let wrap = 2 * l + 2 * w;
  let bow = l * w * h;
  let ribbon = wrap + bow;

  totalPaper += paper;
  totalRibbon += ribbon;

  return;
};

for (let i = 0; i < data.length; i++) {
  // last array elem is a blank string
  if (data[i]) {
    paperAndRibbon(data[i]);
  }
}

console.log(totalPaper); // 1588178
console.log(totalRibbon); // 3783758
